### PR TITLE
DO NOT MERGE: Fixes for Starcoder2 PR

### PR DESCRIPTION
**Pull request by younesbelkada**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -491,6 +491,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
...
 1. **[SqueezeBERT](https://huggingface.co/docs/transformers/model_doc/squeezebert)** (from Berkeley) released with the paper [SqueezeBERT: What can computer vision teach NLP about efficient neural networks?](https://arxiv.org/abs/2006.11316) by Forrest N. Iandola, Albert E. Shaw, Ravi Krishna, and Kurt W. Keutzer.
 1. **[StableLm](https://huggingface.co/docs/transformers/model_doc/stablelm)** (from Stability AI) released with the paper [StableLM 3B 4E1T (Technical Report)](https://stability.wandb.io/stability-llm/stable-lm/reports/StableLM-3B-4E1T--VmlldzoyMjU4?accessToken=u3zujipenkx5g7rtcj9qojjgxpconyjktjkli2po09nffrffdhhchq045vp0wyfo) by Jonathan Tow, Marco Bellagente, Dakota Mahan, Carlos Riquelme Ruiz, Duy Phung, Maksym Zhuravinskyi, Nathan Cooper, Nikhil Pinnaparaju, Reshinth Adithyan, and James Baicoianu.
+1. **[Starcoder2](https://huggingface.co/docs/transformers/main/model_doc/starcoder2)** (from <FILL INSTITUTION>) released with the paper [<FILL PAPER TITLE>](<FILL ARKIV LINK>) by <FILL AUTHORS>.
```
Comment (ArthurZucker) [inline README.md:494]: TOdo

File: docs/source/en/_toctree.yml [hunk 2]
```diff
@@ -132,7 +132,7 @@
   - local: custom_tools
     title: Custom Tools and Prompts
   - local: troubleshooting
-    title: Troubleshoot
```
Comment (ArthurZucker) [inline docs/source/en/_toctree.yml:135]: revert make style probably

File: docs/source/en/model_doc/starcoder2.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<!--Copyright 2024 Mistral AI and The HuggingFace Team. All rights reserved.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/starcoder2.md:1]:
```suggestion
<!--Copyright 2024 The HuggingFace Team. All rights reserved.
```

File: docs/source/en/model_doc/starcoder2.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+-->
+
+# Starcoder2
+
+## Overview
+
+The Starcoder2 model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
+<INSERT SHORT SUMMARY HERE>
+
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/starcoder2.md:23]: TODO as well

File: src/transformers/models/starcoder2/configuration_starcoder2.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+logger = logging.get_logger(__name__)
+
+STARCODER2_PRETRAINED_CONFIG_ARCHIVE_MAP = {}
+
+
+class Starcoder2Config(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`Starcoder2Model`]. It is used to instantiate a
+    Starcoder2 model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the starcoder2-7b_16k model.
```
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/configuration_starcoder2.py:30]:
```suggestion
    with the defaults will yield a similar configuration to that of the [`"starcoder2-7b_16k"`](https://huggingface.co/bigcode/starcoder2-7b_16k) model.
```

File: src/transformers/models/starcoder2/configuration_starcoder2.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    [bigcode/starcoder2-7b_16k](https://huggingface.co/bigcode/starcoder2-7b_16k)
```
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/configuration_starcoder2.py:32]:
```suggestion
```

File: tests/models/starcoder2/test_modeling_starcoder2.py [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+    @require_torch_gpu
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_inference_padding_right(self):
+        self.skipTest("Starcoder2 flash attention does not support right padding")
+
+
+@slow
+@require_torch_gpu
+class Starcoder2IntegrationTest(unittest.TestCase):
```
Comment (ArthurZucker) [inline tests/models/starcoder2/test_modeling_starcoder2.py:470]: let's test eager and flash attention as well! 🤗

File: src/transformers/models/starcoder2/modeling_starcoder2.py [hunk 8]
```diff
@@ -0,0 +1,1377 @@
...
+
+import torch
+import torch.nn.functional as F
+import torch.utils.checkpoint
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+
+from ...activations import ACT2FN
+from ...cache_utils import Cache, DynamicCache
+from ...modeling_attn_mask_utils import _prepare_4d_causal_attention_mask, _prepare_4d_causal_attention_mask_for_sdpa
```
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modeling_starcoder2.py:34]: For a followup-pr: we should not use these attention function and should rely on `Llama`

File: src/transformers/models/starcoder2/modeling_starcoder2.py [hunk 9]
```diff
@@ -0,0 +1,1377 @@
...
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # retrieve input_ids and inputs_embeds
+        if input_ids is not None and inputs_embeds is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/starcoder2/modeling_starcoder2.py:942]: TODO as well when propagating the changes for static kv cache cc @zucchini-nlp and @gante for your TODOs!

**Review by ArthurZucker — APPROVED** [review]
LGTM! 🤗

File: docs/source/en/_toctree.yml (hunk 2 above)
Comment (younesbelkada) [reply docs/source/en/_toctree.yml]: Done!

File: tests/models/starcoder2/test_modeling_starcoder2.py (hunk 7 above)
Comment (younesbelkada) [reply tests/models/starcoder2/test_modeling_starcoder2.py:470]: Done!

**Pull request closed without merging**